### ISSUE TITLE
IProperties::getProperties value not automatically converted with elementMap

### ISSUE THREAD

**Issue by PVince81** [fruux/sabre-dav#767]
### Steps:

1. Create a node that implements IProperties
1. Implement `getProperties` and make it return a `DateTime` (or any other custom class)
1. Implement a Sabre plugin that maps `DateTime` to a serializer class using `$server->xml->elementMap`

### Expected result:
When retrieving the properties through `getProperties`, I'd expect the server to pass the non-string values to element map handlers

### Actual result
Values not converted, serializer complains:
"The writer cannot serialize objects of class: DateTime"

The workaround is to use a `$this->xml->classMap` instead.

Now if you say that this is by design, then the documentation of the IProperties::getProperties() needs to be adjusted to clearly specify that all values must be strings / already serialized strings.

@evert @blizzz

**Comment by evert** [comment]
You're right that the `elementMap` goes in a single direction. The original intent was that the elementMap is used to do:

    XML element -> PHP object

The reason the opposite mapping did not exist, is because PHP objects can implement `Sabre\Xml\XmlSerializable` and be responsible for serializing themselves.

So the easiest way to do this, is by wrapping your PHP `DateTime` object into a small php class looking something like this:

```php

class MyDateTime implements \Sabre\Xml\XmlSerializable {

    protected $value;

    function __construct(\DateTime $value) {
        $this->value = $value;
    }

    function xmlSerialize(\Sabre\Xml\Writer $writer) {
        $writer->write(
            $this->value->format(\DateTime::ATOM)
        );
    }
}

```

This is how nearly all of the existing complex sabre/dav webdav properties behave.

However, since sabre/xml 1.3 we added the `classMap` feature that also does what you describe. To use that, you'd have to call:

```php
$server->xml->classMap['DateTime'] = function(\Sabre\Xml\Writer $writer) {
        $writer->write(
            $this->value->format(\DateTime::ATOM)
        );
};
```

The big difference is that by doing this, you basically create a new global behavior for every PHP `DateTime` object, and I'm not sure how desirable that is.

**Comment by evert** [comment]
I'm leaving this ticket open as something that might need to be clarified though.
